// typechecker/es5/lib/typechecker.js
// Define
'use strict';

var types = ['Array', 'Boolean', 'Date', 'Error', 'Class', 'Function', 'Null', 'Number', 'RegExp', 'String', 'Undefined', 'Object' // deliberately last, as this is a catch all
];

var typeChecker = {

	// -----------------------------------
	// Helpers

	// Get the object type string
	getObjectType: function getObjectType(value) {
		return Object.prototype.toString.call(value);
	},

	// Get the type
	getType: function getType(value) {
		// Cycle
		var _iteratorNormalCompletion = true;
		var _didIteratorError = false;
		var _iteratorError = undefined;

		try {
			for (var _iterator = types[Symbol.iterator](), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {
				var type = _step.value;

				if (typeChecker['is' + type](value)) {
					return type.toLowerCase();
				}
			}

			// Return
		} catch (err) {
			_didIteratorError = true;
			_iteratorError = err;
		} finally {
			try {
				if (!_iteratorNormalCompletion && _iterator['return']) {
					_iterator['return']();
				}
			} finally {
				if (_didIteratorError) {
					throw _iteratorError;
				}
			}
		}

		return null;
	},

	// -----------------------------------
	// Values

	// Checks to see if a value is an object and only an object
	isPlainObject: function isPlainObject(value) {
		/* eslint no-proto:0 */
		return typeChecker.isObject(value) && value.__proto__ === Object.prototype;
	},

	// Checks to see if a value is empty
	isEmpty: function isEmpty(value) {
		return value == null;
	},

	// Is empty object
	isEmptyObject: function isEmptyObject(value) {
		// We could use Object.keys, but this is more effecient
		for (var key in value) {
			if (value.hasOwnProperty(key)) {
				return false;
			}
		}
		return true;
	},

	// Is ES6+ class
	// If changed, isClass must also be updated
	isNativeClass: function isNativeClass(value) {
		return typeof value === 'function' && value.toString().indexOf('class') === 0;
	},

	// Is Conventional Class
	// Looks for function with capital first letter: function MyClass
	// First letter is the 9th character
	// Uppercase letters are between 65 and 90 inclusive
	// If changed, isClass must also be updated
	isConventionalClass: function isConventionalClass(value) {
		var c = undefined;return typeof value === 'function' && (c = value.toString().charCodeAt(9)) >= 65 && c <= 90;
	},

	// There use to be code here that checked for CoffeeScript's "function _Class" at index 0 (which was sound)
	// But it would also check for Babel's __classCallCheck anywhere in the function, which wasn't sound
	// as somewhere in the function, another class could be defined, which would provide a false positive
	// So instead, proxied classes are ignored, as we can't guarantee their accuracy, would also be an ever growing set

	// -----------------------------------
	// Types

	// Is Class
	isClass: function isClass(value) {
		var s = undefined,
		    c = undefined;return typeof value === 'function' && ((s = value.toString()).indexOf('class') === 0 || (c = s.charCodeAt(9)) >= 65 && c <= 90);
	},

	// Checks to see if a value is an object
	isObject: function isObject(value) {
		// null and undefined are objects, hence the truthy check
		return value && typeof value === 'object';
	},

	// Checks to see if a value is an error
	isError: function isError(value) {
		return value instanceof Error;
	},

	// Checks to see if a value is a date
	isDate: function isDate(value) {
		return typeChecker.getObjectType(value) === '[object Date]';
	},

	// Checks to see if a value is an arguments object
	isArguments: function isArguments(value) {
		return typeChecker.getObjectType(value) === '[object Arguments]';
	},

	// Checks to see if a value is a function
	isFunction: function isFunction(value) {
		return typeChecker.getObjectType(value) === '[object Function]';
	},

	// Checks to see if a value is an regex
	isRegExp: function isRegExp(value) {
		return typeChecker.getObjectType(value) === '[object RegExp]';
	},

	// Checks to see if a value is an array
	isArray: function isArray(value) {
		return Array.isArray && Array.isArray(value) || typeChecker.getObjectType(value) === '[object Array]';
	},

	// Checks to see if a valule is a number
	isNumber: function isNumber(value) {
		return typeof value === 'number' || typeChecker.getObjectType(value) === '[object Number]';
	},

	// Checks to see if a value is a string
	isString: function isString(value) {
		return typeof value === 'string' || typeChecker.getObjectType(value) === '[object String]';
	},

	// Checks to see if a valule is a boolean
	isBoolean: function isBoolean(value) {
		return value === true || value === false || typeChecker.getObjectType(value) === '[object Boolean]';
	},

	// Checks to see if a value is null
	isNull: function isNull(value) {
		return value === null;
	},

	// Checks to see if a value is undefined
	isUndefined: function isUndefined(value) {
		return typeof value === 'undefined';
	}
};

// Export
module.exports = typeChecker;
